test/stream_group_test.dart

```diff
@@ -734,6 +734,107 @@ void regardlessOfType(StreamGroup<String> Function() newStreamGroup) {
       expect(events, equals(['one', 'two', 'three', 'four', 'five', 'six']));
     });
   });
+
+  group('onIdle', () {
+    test('emits an event when the last pending stream emits done', () async {
+      streamGroup.stream.listen(null);
+
+      var idle = false;
+      streamGroup.onIdle.listen((_) => idle = true);
+
+      var controller1 = StreamController<String>();
+      var controller2 = StreamController<String>();
+      var controller3 = StreamController<String>();
+
+      streamGroup.add(controller1.stream);
+      streamGroup.add(controller2.stream);
+      streamGroup.add(controller3.stream);
+
+      await flushMicrotasks();
+      expect(idle, isFalse);
+      expect(streamGroup.isIdle, isFalse);
+
+      controller1.close();
+      await flushMicrotasks();
+      expect(idle, isFalse);
+      expect(streamGroup.isIdle, isFalse);
+
+      controller2.close();
+      await flushMicrotasks();
+      expect(idle, isFalse);
+      expect(streamGroup.isIdle, isFalse);
+
+      controller3.close();
+      await flushMicrotasks();
+      expect(idle, isTrue);
+      expect(streamGroup.isIdle, isTrue);
+    });
+
+    test('emits an event each time it becomes idle', () async {
+      streamGroup.stream.listen(null);
+
+      var idle = false;
+      streamGroup.onIdle.listen((_) => idle = true);
+
+      var controller = StreamController<String>();
+      streamGroup.add(controller.stream);
+
+      controller.close();
+      await flushMicrotasks();
+      expect(idle, isTrue);
+      expect(streamGroup.isIdle, isTrue);
+
+      idle = false;
+      controller = StreamController<String>();
+      streamGroup.add(controller.stream);
+
+      await flushMicrotasks();
+      expect(idle, isFalse);
+      expect(streamGroup.isIdle, isFalse);
+
+      controller.close();
+      await flushMicrotasks();
+      expect(idle, isTrue);
+      expect(streamGroup.isIdle, isTrue);
+    });
+
+    test('emits an event when the group closes', () async {
+      // It's important that the order of events here stays consistent over
+      // time, since code may rely on it in subtle ways.
```

Comment:
I don't think we should have any enforced ordering between the stream done and the onIdle done.

cc @lrhn for thoughts on testing behavior that we don't want to make guarantees about.

Reply:
I think there's a difference between "publicly making a guarantee that the behavior will remain consistent", and "internally trying to ensure that the behavior will remain consistent". I agree that you shouldn't advertise this behavior, or consider it part of the public API; however, I think it's worth doing what you can to avoid breaking it *accidentally* and thereby risking a bunch of subtle bugs without any benefit.

Reply:
Honest answer: I'd rather randomize things that people shouldn't depend on, so they can learn to not depend on them. It's a pain to change *anything* timing related because people have written bad code depending on the implementation, not the specification.

In practice, I'm fine testing that the implementation does what we expect the implementation to do, as long as it's *clearly documented* that a change of behavior is fine as long as you also change the test. It's often easier to test the actual behavior than test "any of a number of correct behaviors", you'll likelyend up missing some technically correct behaviors anyway.

Reply:
Added more explicit documentation.